src/services/nnas/routes/oauth.ts

```diff
@@ -20,110 +20,112 @@ router.post('/access_token/generate', async (request: express.Request, response:
 	const refreshToken = request.body.refresh_token;
 
 	if (!['password', 'refresh_token'].includes(grantType)) {
-		response.status(400).send(xmlbuilder.create({
-			error: {
-				cause: 'grant_type',
-				code: '0004',
-				message: 'Invalid Grant Type'
-			}
-		}).end());
-
-		return;
+		return createNNASErrorResponse(response, {
+			errors: [
+				{
+					cause: 'grant_type',
+					code: '0004',
+					message: 'Invalid Grant Type'
+				}
+			]
+		});
 	}
 
 	let pnid = null;
 
 	if (grantType === 'password') {
 		if (!username || username.trim() === '') {
-			response.status(400).send(xmlbuilder.create({
-				error: {
-					cause: 'user_id',
-					code: '0002',
-					message: 'user_id format is invalid'
-				}
-			}).end());
-
-			return;
+			return createNNASErrorResponse(response, {
+				errors: [
+					{
+						cause: 'user_id',
+						code: '0002',
+						message: 'user_id format is invalid'
+					}
+				]
+			});
 		}
 
 		if (!password || password.trim() === '') {
-			response.status(400).send(xmlbuilder.create({
-				error: {
-					cause: 'password',
-					code: '0002',
-					message: 'password format is invalid'
-				}
-			}).end());
-
-			return;
+			return createNNASErrorResponse(response, {
+				errors: [
+					{
+						cause: 'password',
+						code: '0002',
+						message: 'password format is invalid'
+					}
+				]
+			});
 		}
 
 		pnid = await getPNIDByUsername(username);
 
+		// TODO - Client also sends a password_type field, which is always set to "hashed". If field is missing or not "hashed", assume input password is plain-text
 		if (!pnid || !await bcrypt.compare(password, pnid.password)) {
-			response.status(400).send(xmlbuilder.create({
-				errors: {
-					error: {
+			return createNNASErrorResponse(response, {
+				errors: [
+					{
 						code: '0106',
 						message: 'Invalid account ID or password'
 					}
-				}
-			}).end({ pretty: true }));
-
-			return;
+				]
+			});
 		}
 	} else {
 		if (!refreshToken || refreshToken.trim() === '') {
-			response.status(400).send(xmlbuilder.create({
-				error: {
-					cause: 'refresh_token',
-					code: '0106',
-					message: 'Invalid Refresh Token'
-				}
-			}).end());
-
-			return;
+			return createNNASErrorResponse(response, {
+				status: 401,
+				errors: [
+					{
+						cause: 'refresh_token',
+						code: '0106',
+						message: 'Invalid Refresh Token'
+					}
+				]
+			});

```

Comment:
All the `refresh_token` errors in this file have been changed from `400` to `401`.
Was that intentional? nowhere else you seem to have modified status codes.

Reply:
It was. Those are the status codes Nintendo sends, the other status codes were wrong

I was going to update ALL the errors across NNAS, but I still have a lot of research to do on that front and it's not critical to this PRs scope

Reply:
Makes sense, was just making sure